### PR TITLE
Support null value fields in generic row ser/de

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
- Replace `GenericRowSerDeUtils` with `GenericRowSerializer` and `GenericRowDeserailizer` for the following enhancement:
  - Support null value fields ser/de to preserve the null fields information
  - Cache the encoded string bytes to avoid encoding strings twice
  - Support partial deserialize values to only deserialize the needed fields when sorting the rows
- Add `SortOrderComparator` to compare the partial deserialized values
- Modify `ConcatCollector` to use the new classes
- Add `GenericRowSerDeTest` to test the new classes

### CODE REVIEW

**Comment by yupeng9** [comment]
For null value support, does this still require enabling null via config?

**Comment by Jackie-Jiang** [comment]
> For null value support, does this still require enabling null via config?

Yes. The `nullValueFields` are set in `GenericRow` only when the null value is enabled in table config. The change in this PR only make ser/de preserve the `nullValueFields` info.

**Comment by yupeng9** [comment]
> > For null value support, does this still require enabling null via config?
> 
> Yes. The `nullValueFields` are set in `GenericRow` only when the null value is enabled in table config. The change in this PR only make ser/de preserve the `nullValueFields` info.

If the ser/de preserve the null info, then do we still need the enabling config? Or shall we just always enable null handling?

**Comment by Jackie-Jiang** [comment]
> > > For null value support, does this still require enabling null via config?
> > 
> > 
> > Yes. The `nullValueFields` are set in `GenericRow` only when the null value is enabled in table config. The change in this PR only make ser/de preserve the `nullValueFields` info.
> 
> If the ser/de preserve the null info, then do we still need the enabling config? Or shall we just always enable null handling?

Yes, you still need to enable the null handling in the table config

File: pinot-core/src/main/java/org/apache/pinot/core/segment/processing/serde/GenericRowSerializer.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            byteBuffer.putInt(stringBytes.length);
+            byteBuffer.put(stringBytes);
+            break;
+          case BYTES:
+            byte[] bytes = (byte[]) value;
+            byteBuffer.putInt(bytes.length);
+            byteBuffer.put(bytes);
+            break;
+          default:
+            throw new IllegalStateException();
```
Comment (npawar) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/processing/serde/GenericRowSerializer.java:177]: nit: even though this default block will never be reached, add message to the exception? for code readability and also to catch any mistakes made by future development

File: pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/SortOrderComparator.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          result = Double.compare((double) value1, (double) value2);
+          break;
+        case STRING:
+          result = ((String) value1).compareTo((String) value2);
+          break;
+        case BYTES:
+          result = ByteArray.compare((byte[]) value1, (byte[]) value2);
+          break;
+        default:
+          throw new IllegalStateException();
```
Comment (npawar) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/processing/collector/SortOrderComparator.java:64]: exception msg here? would be useful if this is used in places other than the ConcatCollector too

**Review by npawar — APPROVED** [review]
Looks good to me

**Pull request merged**